File: sdks/node/test/unit/errors.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { ERROR_TAXONOMY, FireweaveError, redactSecrets } from '@fireweaveai/server-sdk';

test('taxonomy has exactly the 15 canonical kinds', () => {
  const kinds = Object.keys(ERROR_TAXONOMY).sort();
  assert.deepEqual(kinds, [
    'AlreadyClosed',
    'Authentication',
    'Authorization',
    'BackendUnavailable',
    'Configuration',
    'FlagNotFound',
    'Internal',
    'InvalidContext',
    'MalformedResponse',
    'Network',
    'NotReady',
    'RateLimited',
    'Timeout',
    'TypeMismatch',
    'UnsupportedCapability',
  ]);
  assert.equal(kinds.length, 15);
});

test('taxonomy matches contracts/errors.json mappings', async () => {
  const { readFileSync } = await import('node:fs');
  const { join, dirname } = await import('node:path');
  const { fileURLToPath } = await import('node:url');
  const here = dirname(fileURLToPath(import.meta.url));
  const contractPath = join(here, '..', '..', '..', '..', 'contracts', 'errors.json');
  const contract = JSON.parse(readFileSync(contractPath, 'utf8')) as {
    errors: Array<{ kind: string; openFeatureErrorCode: string; retryable: boolean; defaultMessage: string }>;
  };
  for (const entry of contract.errors) {
    const spec = ERROR_TAXONOMY[entry.kind as keyof typeof ERROR_TAXONOMY];
    assert.ok(spec !== undefined, `missing kind ${entry.kind}`);
    assert.equal(spec.openFeatureErrorCode, entry.openFeatureErrorCode, entry.kind);
    assert.equal(spec.retryable, entry.retryable, entry.kind);
    assert.equal(spec.defaultMessage, entry.defaultMessage, entry.kind);
  }
});

test('AlreadyClosed maps to PROVIDER_NOT_READY', () => {
  assert.equal(ERROR_TAXONOMY.AlreadyClosed.openFeatureErrorCode, 'PROVIDER_NOT_READY');
});

test('redactSecrets strips API keys and bearer tokens', () => {
  const input = 'auth failed for phc_SUPERSECRET123 with Bearer abc.def.ghi and phs_topsecret';
  const out = redactSecrets(input);
  assert.ok(!out.includes('phc_'), out);
  assert.ok(!out.includes('SUPERSECRET'), out);
  assert.ok(!out.includes('abc.def.ghi'), out);
  assert.ok(!out.includes('phs_'), out);
});

test('FireweaveError preserves cause internally, redacts message', () => {
  const cause = new Error('backend said phc_ABC123 is invalid');
  const err = new FireweaveError('Authentication', { message: cause.message, cause });
  assert.equal(err.cause, cause);
  assert.ok(!err.message.includes('phc_ABC123'));
  assert.equal(err.safeMessage, 'authentication failed');
  assert.equal(err.kind, 'Authentication');
  assert.equal(err.openFeatureErrorCode, 'GENERAL');
});

test('default messages carry no secrets and are deterministic', () => {
  for (const spec of Object.values(ERROR_TAXONOMY)) {
    assert.ok(!/ph[csx]_/.test(spec.defaultMessage));
    assert.equal(new FireweaveError(spec.kind).message, spec.defaultMessage);
  }
});

File: sdks/dart/test/errors_test.dart
import 'package:fireweave/fireweave.dart';
import 'package:test/test.dart';

void main() {
  group('redactSecrets', () {
    test('redacts project key prefixes', () {
      expect(
        redactSecrets('key phc_SUPERSECRET0000 leaked'),
        'key [REDACTED] leaked',
      );
      expect(redactSecrets('phs_abc-DEF_123'), '[REDACTED]');
      expect(redactSecrets('phx_'), '[REDACTED]');
    });

    test('redacts bearer tokens', () {
      expect(
        redactSecrets('Authorization: Bearer abc.def.ghi'),
        'Authorization: [REDACTED]',
      );
    });

    test('redacts FW_PROJECT_API_KEY assignments', () {
      expect(redactSecrets('FW_PROJECT_API_KEY=supersecret'), '[REDACTED]');
      expect(redactSecrets('FW_PROJECT_API_KEY : supersecret'), '[REDACTED]');
      // No assignment marker -> not matched (mirrors the reference regex).
      expect(
        redactSecrets('FW_PROJECT_API_KEY is unset'),
        'FW_PROJECT_API_KEY is unset',
      );
    });

    test('collapses whitespace and trims', () {
      expect(redactSecrets('  a   b\n\tc  '), 'a b c');
    });

    test('leaves ordinary text alone', () {
      expect(redactSecrets('invalid configuration'), 'invalid configuration');
    });
  });

  group('ErrorKind / FireweaveError', () {
    test('taxonomy has fifteen members', () {
      expect(ErrorKind.values, hasLength(15));
      expect(ErrorKind.fromWireName('FlagNotFound'), ErrorKind.flagNotFound);
      expect(ErrorKind.fromWireName('Nope'), isNull);
    });

    test('targeting key missing overrides the error code', () {
      final err = FireweaveError.targetingKeyMissing();
      expect(err.openFeatureErrorCode, 'TARGETING_KEY_MISSING');
      expect(err.kind, ErrorKind.invalidContext);
      expect(err.message, 'targeting key missing');
    });

    test('configuration initFatal overrides the error code', () {
      expect(
        FireweaveError.configuration(
          'bad host',
          initFatal: true,
        ).openFeatureErrorCode,
        'PROVIDER_FATAL',
      );
      expect(
        FireweaveError.configuration(
          'bad host',
          initFatal: false,
        ).openFeatureErrorCode,
        'GENERAL',
      );
    });

    test('alreadyClosed maps to PROVIDER_NOT_READY', () {
      expect(
        FireweaveError(ErrorKind.alreadyClosed).openFeatureErrorCode,
        'PROVIDER_NOT_READY',
      );
    });

    test('retryable kinds are exactly the documented five', () {
      const retryable = <ErrorKind>{
        ErrorKind.notReady,
        ErrorKind.rateLimited,
        ErrorKind.timeout,
        ErrorKind.network,
        ErrorKind.backendUnavailable,
      };
      for (final kind in ErrorKind.values) {
        expect(
          kind.isRetryable,
          retryable.contains(kind),
          reason: kind.wireName,
        );
      }
    });

    test('messages are redacted at construction', () {
      final err = FireweaveError(
        ErrorKind.authentication,
        message: 'rejected key phc_LEAK',
      );
      expect(err.message, 'rejected key [REDACTED]');
      expect(FireweaveError(ErrorKind.flagNotFound).message, 'flag not found');
    });
  });
}
